lib/src/rules/use_late.dart

```diff
@@ -0,0 +1,150 @@
+  UseLate()
+      : super(
+          name: 'use_late',
+          description: _desc,
+          details: _details,
+          maturity: Maturity.experimental,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends UnifyingAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  final lateables = <VariableDeclaration>[];
+  final nullableAccess = <AstNode, Element>{};
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      super.visitCompilationUnit(node);
+      _checkAccess();
+    }
+  }
+
+  @override
+  void visitNode(AstNode node) {
+    var element = DartTypeUtilities.getCanonicalElementFromIdentifier(node);
+    if (element != null) {
+      var parent = node.parent;
+      if (parent is Expression) {
+        parent = (parent as Expression).unParenthesized;
+      }
+      if (parent is PostfixExpression &&
+          parent.operand == node &&
+          parent.operator.type == TokenType.BANG) {
+        // ok non-null access
+      } else if (parent is AssignmentExpression &&
+          parent.operator.type == TokenType.EQ &&
+          context.typeSystem.isNonNullable(parent.rightHandSide.staticType)) {
+        // ok non-null access
+      } else {
+        nullableAccess[node] = element;
+      }
+    }
+    super.visitNode(node);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    for (var variable in node.fields.variables) {
+      final parent = node.parent;
+      if (Identifier.isPrivateName(variable.name.name) ||
+          _isPrivateNamedCompilationUnitMember(parent) ||
+          _isPrivateExtension(parent)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  @override
+  void visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
+    for (var variable in node.variables.variables) {
+      if (Identifier.isPrivateName(variable.name.name)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  void _visit(VariableDeclaration variable) {
+    if (variable.isLate) {
+      return;
+    }
+    if (variable.isSynthetic) {
+      return;
+    }
+    if (context.typeSystem.isNonNullable(variable.declaredElement.type)) {
+      return;
+    }
+    lateables.add(variable);
+  }
+
+  void _checkAccess() {
+    for (var variable in lateables) {
+      final lateable = nullableAccess.entries
+          .where((e) =>
+              e.value == variable.declaredElement &&
+              e.key.offset != variable.offset)
+          .isEmpty;
+      if (lateable) {
+        rule.reportLint(variable);
+      }
+    }
+  }
+}
+
+bool _isPrivateNamedCompilationUnitMember(AstNode parent) =>
```

Comment:
Per Erik's comments, if we want to support public fields in private classes, we need to also ensure that there are no instances of either the enclosing class or any subclass of the enclosing class that are ever accessible outside this library.

I suspect that it isn't worth the effort to include these cases, at least at this point, but either way it would be worth capturing this idea and the requirements for doing so in a comment somewhere in this file.

Reply:
Comment added